lualib/rspamadm/clickhouse.lua

```diff
@@ -0,0 +1,247 @@
+    local is_spam = true
+    if r['Action'] == 'no action' or r['Action'] == 'greylist' then
+      is_spam = false
+    end
+    seen_total = seen_total + 1
+    for _, sym in ipairs(r['Symbols.Names']) do
+      local t = known_symbols[sym]
+      if not t then
+        local spam_count, ham_count = 0, 0
+        if is_spam then
+          spam_count = spam_count + 1
+        else
+          ham_count = ham_count + 1
+        end
+        known_symbols[sym] = {
+          id = symbols_count,
+	  seen = 1,
+	  seen_ham = ham_count,
+	  seen_spam = spam_count,
+        }
+	symbols_count = symbols_count + 1
+      else
+        known_symbols[sym].seen = known_symbols[sym].seen + 1
+	if is_spam then
+          known_symbols[sym].seen_spam = known_symbols[sym].seen_spam + 1
+        else
+          known_symbols[sym].seen_ham = known_symbols[sym].seen_ham + 1
+        end
+      end
+    end
+  end
+
+  local known_symbols_list = {}
+  for k, v in pairs(known_symbols) do
+    local lower_count, higher_count
+    if v.seen_spam > v.seen_ham then
+      lower_count = v.seen_ham
+      higher_count = v.seen_spam
+    else
+      lower_count = v.seen_spam
+      higher_count = v.seen_ham
+    end
+    if lower_count / higher_count >= 0.95 then
+      -- Weak ham/spam correlation, remove symbol
+      print(string.format("Removing %s - weak ham/spam correlation", k))
+      known_symbols[k] = nil
+      removed = removed + 1
+    elseif v.seen / seen_total >= 0.9 then
+      -- Symbol is omnipresent, remove it
+      print(string.format("Removing %s - omnipresent symbol", k))
+      known_symbols[k] = nil
+      removed = removed + 1
+    else
+      known_symbols_list[v.id] = {
+        seen = v.seen,
+        name = k,
+      }
+    end
+  end
+
+  -- Walk results again & count correlations
+  local correlations = {}
+  for _, r in ipairs(res) do
+    for _, sym in ipairs(r['Symbols.Names']) do
+      for _, inner_sym_name in ipairs(r['Symbols.Names']) do
+        if inner_sym_name ~= sym then
+          local known_sym = known_symbols[sym]
+          local inner_sym = known_symbols[inner_sym_name]
+          if known_sym and inner_sym then
+            if not correlations[known_sym.id] then
+              correlations[known_sym.id] = {}
+            end
+            local n = correlations[known_sym.id][inner_sym.id] or 0
+            n = n + 1
+            correlations[known_sym.id][inner_sym.id] = n
+          end
+	end
+      end
+    end
+  end
+
+  -- Walk correlation matrix and divide by total counts
+  local remove = {}
+  for sym_id, row in pairs(correlations) do
+    for inner_sym_id, count in pairs(row) do
+      local known = known_symbols_list[sym_id]
+      if known and count == known.seen then
+        print(string.format("Removing %s - overlapped with %s",
+            known.name, known_symbols_list[inner_sym_id].name))
+        remove[known.name] = true
+      end
+    end
+  end
+
+  for k in pairs(remove) do
```

Comment:
Why do we need to remove it from the table? Just keep two tables and when iterating on `know_symbols` check for `remove` and add a corresponding message. We by the way want to have both symbols that are excluded and symbols that are included + json/ucl format to export those directly to a config file.